src/tensor_operations/matrix_decomposition.jl

```diff
@@ -360,46 +360,174 @@ function noinds_error_message(decomp::String)
+  lre = isempty(Lis) || isempty(Ris)
+  # make a dummy index with dim=1 and incorporate into A so the Lis & Ris can never
+  # be empty.  A essentially becomes 1D after collection.
+  if (lre)
+    A, vαl, vαr, Lis, Ris = add_trivial_index(A, Lis, Ris)
+  end
 
-  Lis_original = Lis
-  Ris_original = Ris
-  if isempty(Lis_original)
-    α = trivial_index(Ris)
-    vLα = onehot(eltype(A), α => 1)
-    A *= vLα
-    Lis = [α]
+  #
+  #  Use combiners to render A down to a rank 2 tensor ready matrix QR routine.
+  #
+  CL, CR = combiner(Lis...), combiner(Ris...)
+  cL, cR = combinedind(CL), combinedind(CR)
+  AC = A * CR * CL
+  #
+  #  Make sure we don't accidentally pass the transpose into the matrix qr routine.
+  #
+  if inds(AC) != IndexSet(cL, cR)
+    AC = permute(AC, cL, cR)
   end
-  if isempty(Ris_original)
-    α = trivial_index(Lis)
-    vRα = onehot(eltype(A), α => 1)
-    A *= vRα
-    Ris = [α]
+  # qr the matrix.
+  QT, RT = qr(tensor(AC); kwargs...)
+
+  # We need a use case that fails without this code.
+  # correcting the fluxes of the two tensors, such that Q has 0 flux for all blocks
+  # and R has the total flux of the system
+  # if hasqns(AC)
+  # for b in nzblocks(QT)
+  #     i1 = inds(QT)[1]
+  #     i2 = inds(QT)[2]
+  #     r1 = inds(RT)[1]
+  #     newqn = -dir(i2) * flux(i1 => Block(b[1]))
+  #     ITensors.setblockqn!(i2, newqn, b[2])
+  #     ITensors.setblockqn!(r1, newqn, b[2])
+  # end
+  # end
+
+  #
+  #  Undo the combine oepration, to recover all tensor indices.
+  #
+  Q, R = itensor(QT) * dag(CL), itensor(RT) * dag(CR)
+
+  # Conditionally remove dummy indices.
+  if (lre)
+    Q, R = remove_trivial_index(Q, R, vαl, vαr)
   end
-
-  Lpos, Rpos = NDTensors.getperms(inds(A), Lis, Ris)
-  QT, RT = qr(tensor(A), Lpos, Rpos; kwargs...)
-  Q, R = itensor(QT), itensor(RT)
+  #
+  # fix up the tag name for the index between Q and R.
+  #  
   q = commonind(Q, R)
-  settags!(Q, tags, q)
-  settags!(R, tags, q)
-  q = settags(q, tags)
+  settags!(Q, qtag, q)
+  settags!(R, qtag, q)
+  q = settags(q, qtag)
 
-  if isempty(Lis_original)
-    Q *= dag(vLα)
+  return Q, R, q
+end
+
+function rq(A::ITensor, Linds...; kwargs...)
+  qtag::TagSet = get(kwargs, :tags, "Link,rq") #tag for new index between Q and R
+  Lis = commoninds(A, indices(Linds...))
+  Ris = uniqueinds(A, Lis)
+  lre = isempty(Lis) || isempty(Ris)
+  # make a dummy index with dim=1 and incorporate into A so the Lis & Ris can never
+  # be empty.  A essentially becomes 1D after collection.
+  if (lre)
+    A, vαl, vαr, Lis, Ris = add_trivial_index(A, Lis, Ris)
   end
-  if isempty(Ris_original)
-    R *= dag(vRα)
+
+  #
+  #  Use combiners to render A down to a rank 2 tensor ready matrix QR routine.
+  #
+  CL, CR = combiner(Lis...), combiner(Ris...)
+  cL, cR = combinedind(CL), combinedind(CR)
+  AC = A * CR * CL
+  #
+  #  Make sure we don't accidentally pass the transpose into the matrix qr routine.
+  #
+  if inds(AC) != IndexSet(cL, cR)
```

Comment:
`IndexSet` is deprecated. I think this should be: `inds(AC) != (cL, cR)`.

Reply:
Good point.  Do we even need the if statement?  Presumably calling permute on properly order indices would be fast anyway.

Reply:
Good point. The purpose of that check was to avoid unnecessary copying in `permute`, but you can accomplish the same thing by setting `permute(A, inds; allow_alias=true)` (https://itensor.github.io/ITensors.jl/stable/ITensorType.html#ITensors.permute-Tuple{ITensor,%20Any}) to avoid copying when possible.

By default, even if the permutation of the indices is trivial, `permute` will copy the data to avoid any surprising behavior if code modifies the tensor data later on.

Reply:
OK fixed using allow_alias=true

Reply:
Thanks for fixing the CI tests.